[SPARK-26248][SQL] Infer date type from CSV

MaxGekk opened the pull request:
## What changes were proposed in this pull request?

The `CSVInferSchema` class is extended to support inferring of `DateType` from CSV input. The attempt to infer `DateType` is performed after inferring `TimestampType`.

## How was this patch tested?

Added new test for inferring date types from CSV . It was also tested by existing suites like `CSVInferSchemaSuite`, `CsvExpressionsSuite`, `CsvFunctionsSuite` and `CsvSuite`.

MaxGekk committed:
Test for inferring the date type

MaxGekk committed:
Inferring date type

AmplabJenkins commented:
Can one of the admins verify this patch?

SparkQA commented:
**[Test build #99581 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/99581/testReport)** for PR 23202 at commit [`fa915fd`](https://github.com/apache/spark/commit/fa915fd3fcab37be2c53bc23a77b55e3024b3994).

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/testing-k8s-prb-make-spark-distribution-unified/5639/
Test PASSed.

SparkQA commented:
**[Test build #99581 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/99581/testReport)** for PR 23202 at commit [`fa915fd`](https://github.com/apache/spark/commit/fa915fd3fcab37be2c53bc23a77b55e3024b3994).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/99581/
Test PASSed.

HyukjinKwon commented on sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/csv/CSVInferSchema.scala:
```diff
@@ -98,6 +100,7 @@ class CSVInferSchema(options: CSVOptions) extends Serializable {
           compatibleType(typeSoFar, tryParseDecimal(field)).getOrElse(StringType)
         case DoubleType => tryParseDouble(field)
         case TimestampType => tryParseTimestamp(field)
+        case DateType => tryParseDate(field)
```
The problem here is it looks a bit odd that we try date type later. IIRC the root cause is related with date paring library. Couldn't we try date first if we switch the parsing library? I thought that's in progress.

MaxGekk replied:
Just in case, I did exact match here as well, see https://github.com/apache/spark/pull/23202/files#diff-17719da188b2c15129f848f654a0e6feR174 . If date parser didn't consume all input (`pos.getIndex != field.length`), it fails. If I move it up in type inferring pipeline, it should work.

HyukjinKwon replied:
I see. Can we try date first above? was wondering if there was a reason to try date first.

MaxGekk replied:
Done. Please, have a look at the changes.

MaxGekk committed:
Additional tests

MaxGekk committed:
Infer date type before timestamp type

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/testing-k8s-prb-make-spark-distribution-unified/5663/
Test PASSed.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/99607/
Test PASSed.

srowen reviewed (commented):
I don't know this part well but the change looks reasonable.

MaxGekk commented:
@HyukjinKwon @srowen Is there anything which worries you in the PR?

srowen commented:
I'd defer to @HyukjinKwon ; looks OK in broad strokes but he would know much more about the CSV parsing.

HyukjinKwon commented:
Similar discussion is going on at https://github.com/apache/spark/pull/23201#discussion_r240156871. Let me keep tracking them. Sorry for late response, @MaxGekk

MaxGekk committed:
Merge remote-tracking branch 'origin/master' into csv-date-inferring

MaxGekk committed:
Moving on new date parser

MaxGekk committed:
Infer TimestampType before DateType

SparkQA commented:
**[Test build #100194 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/100194/testReport)** for PR 23202 at commit [`0ec5c76`](https://github.com/apache/spark/commit/0ec5c767103b8a20db60a2f57386b8fd22c712d3).

MaxGekk commented:
I have rebased this branch on the master and as a consequence of that `CsvInferSchema` uses new date/timestamp parser for type inference. Can we continue with this PR since it is used new `Date`/`TimeFormatter` introduced by https://github.com/apache/spark/pull/23150 and probably will be not affected by https://github.com/apache/spark/pull/23196. 

Also I changed order of type inference here. For now `TimestampType` is inferred before `DateType`. /cc @cloud-fan

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/testing-k8s-prb-make-spark-distribution-unified/6168/
Test PASSed.

SparkQA commented:
**[Test build #100194 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/100194/testReport)** for PR 23202 at commit [`0ec5c76`](https://github.com/apache/spark/commit/0ec5c767103b8a20db60a2f57386b8fd22c712d3).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/100194/
Test PASSed.

cloud-fan commented:
retest this please

AmplabJenkins commented:
Build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/testing-k8s-prb-make-spark-distribution-unified/6173/
Test PASSed.

SparkQA commented:
**[Test build #100198 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/100198/testReport)** for PR 23202 at commit [`0ec5c76`](https://github.com/apache/spark/commit/0ec5c767103b8a20db60a2f57386b8fd22c712d3).

MaxGekk committed:
Merge branch 'master' into csv-date-inferring

SparkQA commented:
**[Test build #100201 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/100201/testReport)** for PR 23202 at commit [`ba4a9dc`](https://github.com/apache/spark/commit/ba4a9dca02337204a37984845830cd7c8a121656).

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/testing-k8s-prb-make-spark-distribution-unified/6176/
Test PASSed.

SparkQA commented:
**[Test build #100201 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/100201/testReport)** for PR 23202 at commit [`ba4a9dc`](https://github.com/apache/spark/commit/ba4a9dca02337204a37984845830cd7c8a121656).
 * This patch **fails to build**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/100201/
Test FAILed.

MaxGekk committed:
Fix merge

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/testing-k8s-prb-make-spark-distribution-unified/6178/
Test PASSed.

SparkQA commented:
**[Test build #100203 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/100203/testReport)** for PR 23202 at commit [`beb6912`](https://github.com/apache/spark/commit/beb69120f7b2b16121a6d87e9cf34955f1e1eb8f).

SparkQA commented:
**[Test build #100203 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/100203/testReport)** for PR 23202 at commit [`beb6912`](https://github.com/apache/spark/commit/beb69120f7b2b16121a6d87e9cf34955f1e1eb8f).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/100203/
Test PASSed.

HyukjinKwon commented:
Timestamp should not tried first. It works for default values but don't work when, for instance, other patterns are set.

For instnace:


```scala
Seq("2010_10_10").toDF.write.text("/tmp/bar")

spark.read.option("inferSchema", "true")
  .option("header", "false")
  .option("timestampFormat", "yyyy_MM_dd")
  .csv("/tmp/foo").printSchema()
```

```
root
 |-- _c0: date (nullable = true)
```

```scala
Seq("2010-10-10").toDF.write.text("/tmp/foo")

spark.read.option("inferSchema", "true")
  .option("header", "false")
  .option("timestampFormat", "yyyy-MM-dd")
  .csv("/tmp/foo").printSchema()
```

```
root
 |-- _c0: timestamp (nullable = true)
```

HyukjinKwon commented:
That's why CSV didn't introduced date type yet because the pattern can be arbitrarily set. How did we handle exact match problem here, @MaxGekk? Doesn't it cause such problem when legacy is on?

Also, how do we define the precedence between `dateFormat` and `timestampFormat`?

HyukjinKwon commented:
What I mean by exact match is, for instance, we use `FastDateFormat` (which is compatible with `SimpleDateFormat`):

https://github.com/apache/spark/blob/8a27952cdbf492939d9bda59e2f516f574581636/sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/util/TimestampFormatter.scala#L95

https://github.com/apache/spark/blob/8a27952cdbf492939d9bda59e2f516f574581636/sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/util/TimestampFormatter.scala#L157

How do we handle the case below when legacy is on?

```
scala> import org.apache.commons.lang3.time.FastDateFormat
import org.apache.commons.lang3.time.FastDateFormat

scala> FastDateFormat.getInstance("yyyy-MM").parse("2010-10-10")
res19: java.util.Date = Fri Oct 01 00:00:00 SGT 2010

scala> FastDateFormat.getInstance("yyyy-MM-dd").parse("2010-10-10")
res20: java.util.Date = Sun Oct 10 00:00:00 SGT 2010
```

It's going to introduce some arbitrary behaviours to end users.

HyukjinKwon commented:
Let's don't add date type inference support yet until we get rid of legacy conf. Introducing exact match itself is a breaking change .. Since date type inference support is an additional improvement, we can do it later during minor releases as well.

HyukjinKwon commented on sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/csv/CSVInferSchema.scala:
```diff
@@ -104,6 +108,7 @@ class CSVInferSchema(val options: CSVOptions) extends Serializable {
           compatibleType(typeSoFar, tryParseDecimal(field)).getOrElse(StringType)
         case DoubleType => tryParseDouble(field)
         case TimestampType => tryParseTimestamp(field)
+        case DateType => tryParseDate(field)
```
Another problem here is the order here matters when type is being merged. For instance, date type is inferred first, and then if timestamp is found, it won't detect timestamp type anymore.

cloud-fan replied:
IIRC we decided to follow the order in partition inference and infer timestamp first?

HyukjinKwon replied:
Here's what I mean:

```scala
Seq("2010|10|10", "2010_10_10")
  .toDF.repartition(1).write.mode("overwrite").text("/tmp/foo")
spark.read
  .option("inferSchema", "true")
  .option("header", "false")
  .option("dateFormat", "yyyy|MM|dd")
  .option("timestampFormat", "yyyy_MM_dd").csv("/tmp/foo").printSchema()
```

```
root
 |-- _c0: string (nullable = true)
```

```scala
Seq("2010_10_10", "2010|10|10")
  .toDF.repartition(1).write.mode("overwrite").text("/tmp/foo")
spark.read
  .option("inferSchema", "true")
  .option("header", "false")
  .option("dateFormat", "yyyy|MM|dd")
  .option("timestampFormat", "yyyy_MM_dd").csv("/tmp/foo").printSchema()
```

```
root
 |-- _c0: date (nullable = true)
```

```scala
Seq("2010_10_10")
  .toDF.repartition(1).write.mode("overwrite").text("/tmp/foo")
spark.read
  .option("inferSchema", "true")
  .option("header", "false")
  .option("timestampFormat", "yyyy_MM_dd").csv("/tmp/foo").printSchema()
```

```
root
 |-- _c0: timestamp (nullable = true)
```

```scala
Seq("2010|10|10")
  .toDF.repartition(1).write.mode("overwrite").text("/tmp/foo")
spark.read
  .option("inferSchema", "true")
  .option("header", "false")
  .option("dateFormat", "yyyy|MM|dd").csv("/tmp/foo").printSchema()
```

```
root
 |-- _c0: date (nullable = true)

```

HyukjinKwon replied:
It's okay .. sorry for rushing comments. I realised my comments are hard to read now.

HyukjinKwon commented:
@MaxGekk and @cloud-fan, if I am not completely wrong here, I would like to revert this .. but let me wait for your guys input for a while.

cloud-fan commented:
@HyukjinKwon can you elaborate what's broken after this patch? If the legacy stuff is your only concern, can we disable this feature when legacy is on?

HyukjinKwon commented:
Give me a min .. let me summarise with examples ..

HyukjinKwon commented:
Ah, thank you @cloud-fan. Let me leave a comment for summarising it first here anyway. So that we can see what's problem and what should be addressed when this PR is proposed again.

HyukjinKwon commented:
**Problem 1.**

https://github.com/apache/spark/pull/23202#discussion_r242010787 - I left some examples there.

If there are multiple rows, and the first row is inferred as date type in the same partition,
It will not be able to infer timestamp afterward.


**Problem 2.**

https://github.com/apache/spark/pull/23202#issuecomment-447701620

If legacy is on, we have ambiguity about date/timestamp pattern matching, because they can be arbitrarily set by users.
It does not do the exact match, which means it's not going to distinguish `yyyy-MM` and `yyyy-MM-dd` for input, for instane, `2010-10-10`.

```
  test("") {
    withSQLConf("spark.sql.legacy.timeParser.enabled" -> "true") {
      Seq("2010_10_10").toDF
        .write.mode("overwrite").text("/tmp/foo")

      spark.read.option("inferSchema", "true")
        .option("header", "false")
        .option("dateFormat", "yyyy_MM_dd")
        .option("timestampFormat", "yyyy_MM")
        .csv("/tmp/foo").printSchema()
    }
  }
```

it ends up with

```
root
 |-- _c0: date (nullable = true)
```

We are able to do this only when `spark.sql.legacy.timeParser.enabled` is disabled (by default), however, I was thinking it's going to introduce complexity. 
I was thinking we could do it later when we remove `spark.sql.legacy.timeParser.enabled`. Date type inference isn't super important IMHO becase we infer timestamps.
I would like to talk about this further if anyone thinks differently. If the change isn't complicated then I thought, it should also be okay to go ahead.

**Questions:**

How do we define the precedence between `dateFormat` and `timestampFormat`? (for instance, if the patterns are same, then, does it become timestamp or date?)

HyukjinKwon commented:
I have reverted this as discussed with @cloud-fan above.
